packages/core/mx/contracts/staking/src/lib.rs

```diff
@@ -0,0 +1,336 @@
+            return MultiValue2((MultiValueEncoded::new(), MultiValueEncoded::new()));
+        }
+
+        let mut stakers_list: MultiValueEncoded<Staker<Self::Api>> = MultiValueEncoded::new();
+        for staker in stakers_addresses.clone().into_iter() {
+            let stake = self.stakes(&staker).get();
+            stakers_list.push(stake);
+        }
+
+        MultiValue2((stakers_addresses, stakers_list))
+    }
+
+    /// Withdraw staker tokens once the lock period has passed
+    fn withdraw(&self, staker: &mut Staker<Self::Api>, staker_address: &ManagedAddress) {
+        let block_number = self.blockchain().get_block_nonce();
+        let tokens_to_withdraw = staker.withdraw_tokens(block_number);
+        require!(tokens_to_withdraw > 0, "Stake has no available tokens for withdrawal");
+
+        self.send().direct_esdt(staker_address, &self.staking_token().get(), 0, &tokens_to_withdraw);
+        self.stake_withdrawn_event(staker_address, &tokens_to_withdraw);
+    }
+
+    fn get_or_create_staker(&self, staker_address: &ManagedAddress) -> Staker<Self::Api> {
+        if self.stakes(staker_address).is_empty() {
+            let new_staker = Staker::new();
+            self.stakes(staker_address).set(&new_staker);
+
+            return new_staker;
+        }
+
+        self.stakes(staker_address).get()
+    }
+
+    fn get_escrow_allocation_state(&self, escrow_address: &ManagedAddress) -> AllocationState {
+        if self.allocations(escrow_address).is_empty() {
+            return AllocationState::Null;
+        }
+
+        let allocation = self.allocations(&escrow_address).get();
+        let escrow_status = self.get_status(&escrow_address);
+
+        if allocation.created_at != 0 && allocation.tokens > 0 && escrow_status == EscrowStatus::Pending {
+            return AllocationState::Pending;
+        }
+
+        if allocation.closed_at == 0 {
+            match escrow_status {
+                EscrowStatus::Launched => return AllocationState::Active,
+                EscrowStatus::Complete => return AllocationState::Completed,
+                _ => return AllocationState::Closed,
+            }
+        }
+
+        AllocationState::Closed
+    }
+
+    fn require_only_staker(&self){
+        let caller = self.blockchain().get_caller();
+        let stakers = self.stakers().get();
+        require!(stakers.contains(&caller), "Only stakers can call this function");
+    }
+
+    /// Staking token id
+    #[storage_mapper("staking_token")]
+    fn staking_token(&self) -> SingleValueMapper<TokenIdentifier>;
+
+    /// Minimum amount of tokens a staker needs to stake
+    #[storage_mapper("minimum_stake")]
+    fn minimum_stake(&self) -> SingleValueMapper<BigUint>;
+
+    /// Time in blocks to unstake
+    #[storage_mapper("lock_period")]
+    fn lock_period(&self) -> SingleValueMapper<u64>;
+
+    /// List of stakers
+    #[storage_mapper("stakers")]
+    fn stakers(&self) -> SingleValueMapper<ManagedVec<ManagedAddress>>;
```

Comment:
Why not use a UnorderedSetMapper? It allows for checks, insert & remove elements and is iterable.